index.rst

```diff
@@ -1,25 +1,370 @@
+Messages are sent using TCP/IP.
+
+Both DDS and Kafka support retrieving historical (aka late-joiner) data, and our control system architecture requires this.
+It allows a new node to obtain the current state of the system without issuing commands to the running controllers.
+
+Differences Between DDS and Kafka
+=================================
+
+Kafka is a broker-based system, whereas DDS is distributed, with no central broker.
+
+With DDS each node (e.g. CSC or Jupyter Notebook) discovers the other nodes, using UDP, as the software starts up.
+Reliability is achieved by switching which node acts as “master” if one fails.
+
+With Kafka each node connects to a specified broker and schema registry, in order to read and write messages.
+There is no discovery phase and no need for UDP.
+Reliability and performance can be increased by running multiple brokers in parallel, and by using more than one "partition" for each topic.
+
+Schema Evolution
+----------------
+
+DDS does not support schema evolution, at least as we are using it.
+When we update the schema for a topic it becomes an entirely new topic.
+Internally the topic has has a different name; the topic names DDS sees include a hash that represents the schema
+though we hide this hash from users.
+
+Kafka schemas support basic schema evolution: we may add or delete fields and readers and writers can freely use older or newer schemas.
+Note that even with this type of schema evolution, we will have to reset part of the system if we change the type of a message field.
+
+Licensing
+---------
+
+Kafka is open-source, ADLink OpenSplice is not.
+With Kafka developers can run the full system, with all features, without licensing headaches.
+With ADLink OpenSplice we presently run two very different versions of the software, in order to reduce costs and avoid licensing headaches:
+
+* Developers use the “community” edition, which ADLink no longer offers, and which only supports single-process topic daemons.
+
+* Deployment uses the commercial edition, because it supports shared-memory topic daemons, which work better for large systems.
+  The commerical version is also required to run ADLink's diagnostic tools.
+  Note that the commercial edition has always been newer than the community edition.
+  In the past ADLink kept the community edition one major version behind the commercial edition.
+  But the last community version they served is now two major versions behind the commercial edition.
+
+Commercial support for Kafka is available from Confluent, should we need it.
+
+The Engineering Facilities Database
+-----------------------------------
+
+DDS requires running complex extra software (ts_salkafka, an in-house package) in order to copy all messages from DDS to Kafka, so we can ingest them into the Engineering Facilities Database (EFD).
+If we use Kafka then we can eliminate this extra software.
+
+Performance
```

Comment:
So, one thing I think we are missing is a good estimate of the latency. Here you are basically measuring throughput (how many messages you can write/read). Latency is sort of measuring the time between write and read.. For example, we could have a throughput of ~1kHz, as you have been measuring with Kafka, but if the latency is ~1s it means there's a delay of 1s in delivering those messages, even if we are reading and writing them a lot faster. 

What I would suggest we do is measure latency at different throughputs. So, first we need to have a measurement of when the data is written and then, when it is read. I think we have attributes in the topics that measure that. Then, we would write topics at 1Hz, 3Hz, 5Hz, 10Hz, 22Hz, 47Hz, 100Hz, 216Hz, 465Hz, 1000Hz for like a minute or so, and measure the read/write delay. And do this for both the smaller and larger topics. Then we can get the average and standard deviation and estimate the latency and stability.

If you want to do this as part of a separate ticket, I think it is fine as well.

Reply:
That is already in the table! But the table was so wide that one had to scroll horizontally to see all the columns. And if you missed that, so will other readers. I have rewritten the section to use 2 tables. We'll see if that eliminates the scrolling.